### PR TITLE
🌐 Exporting and importing of PO files

### PR DESCRIPTION
**Pull request by rowanseymour**
Captura de Pantalla 2020-04-01 a la(s) 12 21 58
Captura de Pantalla 2020-04-01 a la(s) 12 32 01
Captura de Pantalla 2020-04-01 a la(s) 12 22 34

### CODE REVIEW

**Comment by rowanseymour** [comment]
Captura de Pantalla 2020-04-02 a la(s) 08 55 44

Captura de Pantalla 2020-04-02 a la(s) 10 11 45

**Comment by nicpottier** [comment]
Do we mark what language a PO file was generated for? Just wondering if we can either derive language without them picking it or warn them if they are picking a language that is different than the PO file. Could see people spacing on setting the language in the drop down and being very sad that all their translations were changed.

**Comment by rowanseymour** [comment]
@nicpottier just wrestling with that right now.. I wonder how POEdit and Transifex feel about 3-char language codes? Currently (because best I could tell all PO files use 2 char codes, en-GB etc).. the language code that gets saved into the PO header is 2-char

**Comment by nicpottier** [comment]
No reason we can't have our own header there the same as we do for flow UUIDs right?

**Comment by rowanseymour** [comment]
True could just add `Language-3: fra` alongside `Language: fr`..

**Comment by rowanseymour** [comment]
Looks like POEdit understood `fra` but replaced it with `fr`

File: temba/flows/views.py [hunk 1]
```diff
@@ -1289,6 +1301,178 @@ def get_gear_links(self):
...
+                                _("Contains translations in %(lang)s which is not a supported translation language."),
+                                params={"lang": po_info.language_name},
+                            )
+
+                return data
+
+        class ConfirmForm(forms.Form):
+            language = forms.ChoiceField(
+                label=_("Language"), help_text=_("Replace flow translations in this language."), required=True
+            )
```
Comment (ericnewcomer) [inline temba/flows/views.py:1421]: Should use `temba-select` for new dropdowns, see example for adding campaign events.
https://github.com/nyaruka/rapidpro/blob/master/temba/campaigns/views.py#L300

**Comment by rowanseymour** [comment]
Now with 2 step import..

Captura de Pantalla 2020-04-03 a la(s) 14 11 50
Captura de Pantalla 2020-04-03 a la(s) 14 12 04

File: static/coffee/temba.coffee [hunk 2]
```diff
@@ -468,17 +468,15 @@ class @Modax extends @ConfirmationModal
             if modal.listeners.onSuccess
               modal.listeners.onSuccess(xhr)
 
+          modal.dismiss()
```
Comment (rowanseymour) [inline static/coffee/temba.coffee:471]: need modal to be dismissed even redirect is to a download - thus the browser doesn't actually navigate away from current page

File: temba/utils/gettext.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import os
```
Comment (rowanseymour) [inline temba/utils/gettext.py:1]: arguably doesn't belong in `utils`.. but not sure where else to put it

File: templates/flows/flow_import_translation.haml [hunk 4]
```diff
@@ -0,0 +1,136 @@
+-extends 'smartmin/update.html'
+-load i18n
+-load smartmin
+-load temba
+
+-block pre-form
+  -blocktrans trimmed
+    Importing a PO translation file will replace any translations flow in the selected language, with any matching
+    translations in the PO file.
```
Comment (ericnewcomer) [inline templates/flows/flow_import_translation.haml:9]: `will replace any translations flow` should probably read `will replace any translations in the flow for the selected language` or similar

File: temba/flows/views.py [hunk 5]
```diff
@@ -1289,6 +1302,190 @@ def get_gear_links(self):
                 links.append(dict(title=_("Previous Editor"), js_class="previous-editor", href="#"))
             return links
 
+    class ExportTranslation(OrgObjPermsMixin, ModalMixin, SmartUpdateView):
+        class Form(forms.Form):
+            language = forms.ChoiceField(
+                required=False,
+                label=_("Language"),
+                choices=[("", "None")],
+                help_text=_("Include translations in this language."),
```
Comment (ericnewcomer) [inline temba/flows/views.py:1311]: `SelectWidget`

File: static/coffee/temba.coffee (hunk 2 above)
Comment (nicpottier) [reply static/coffee/temba.coffee:471]: This feels high risk, lets make sure to test this on staging a bunch with other dialogs before shipping.

File: templates/flows/flow_export_translation.haml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+-extends 'smartmin/form.html'
+-load i18n
+-load smartmin
+-load temba
+
+-block pre-form
+  -blocktrans trimmed
+    Extracts a <a href="https://en.wikipedia.org/wiki/Gettext">gettext</a> PO file from the flow which can later be re-imported with updated translations.
```
Comment (nicpottier) [inline templates/flows/flow_export_translation.haml:8]: for these kind of links I prefer to put the whole `<a>` tag in a `with`.. it isn't going to be translated and makes that translation a lot more sane.

File: templates/flows/flow_import_translation.haml [hunk 7]
```diff
@@ -0,0 +1,136 @@
+-extends 'smartmin/update.html'
+-load i18n
+-load smartmin
+-load temba
+
+-block pre-form
+  -blocktrans trimmed
+    Importing a PO translation file will replace any translations in the flow for the selected language, with any
+    matching translations in the PO file.
```
Comment (nicpottier) [inline templates/flows/flow_import_translation.haml:9]: This feels like a rough sentence to grok, maybe just `Importing a PO file will replace all matching translations in the flow for the selected language.` ?

File: templates/flows/flow_export_translation.haml (hunk 6 above)
Comment (rowanseymour) [reply templates/flows/flow_export_translation.haml:8]: So I had that.. and then thought actually https://**en**.wikipedia.org/wiki/Gettext is the one type of link you should translate? Can change if not

Comment (nicpottier) [reply templates/flows/flow_export_translation.haml:8]: Oh hah, ya, I guess! Didn't catch that. Have some doubts that any translator would do that but agree this is probably right.

**Pull request merged**